hazelcast/src/main/java/com/hazelcast/cp/internal/MetadataRaftGroupManager.java

```diff
@@ -0,0 +1,985 @@
+        addRaftGroup(group);
+
+        logger.info("New " + group.id() + " is created with " + group.members());
+
+        RaftGroupId groupId = group.id();
+        if (group.containsMember(getLocalCPMember())) {
+            raftService.createRaftNode(groupId, group.memberImpls());
+        } else {
+            // Broadcast group-info to non-metadata group members
+            OperationService operationService = nodeEngine.getOperationService();
+            CPGroupInfo metadataGroup = groups.get(metadataGroupId);
+            for (CPMemberInfo member : group.memberImpls()) {
+                if (!metadataGroup.containsMember(member)) {
+                    Operation op = new CreateRaftNodeOp(group.id(), (Collection) group.initialMembers());
+                    operationService.send(op, member.getAddress());
+                }
+            }
+        }
+
+        return groupId;
+    }
+
+    private void addRaftGroup(CPGroupInfo group) {
+        CPGroupId groupId = group.id();
+        checkState(!groups.containsKey(groupId), group + " already exists!");
+        groups.put(groupId, group);
+    }
+
+    private CPGroupInfo getRaftGroupByName(String name) {
+        for (CPGroupInfo group : groups.values()) {
+            if (group.status() != DESTROYED && group.name().equals(name)) {
+                return group;
+            }
+        }
+        return null;
+    }
+
+    public void triggerDestroyRaftGroup(CPGroupId groupId) {
+        checkNotNull(groupId);
+        checkState(membershipChangeContext == null,
+                "Cannot destroy " + groupId + " while there are ongoing CP membership changes!");
+
+        CPGroupInfo group = groups.get(groupId);
+        checkNotNull(group, "No CP group exists for " + groupId + " to destroy!");
+
+        if (group.setDestroying()) {
+            logger.info("Destroying " + groupId);
+        } else if (logger.isFineEnabled()) {
+            logger.fine(groupId + " is already " + group.status());
+        }
+    }
+
+    public void completeDestroyRaftGroups(Set<CPGroupId> groupIds) {
+        checkNotNull(groupIds);
+
+        for (CPGroupId groupId : groupIds) {
+            checkNotNull(groupId);
+
+            CPGroupInfo group = groups.get(groupId);
+            checkNotNull(group, groupId + " does not exist to complete destroy");
+
+            completeDestroyRaftGroup(group);
+        }
+    }
+
+    private void completeDestroyRaftGroup(CPGroupInfo group) {
+        CPGroupId groupId = group.id();
+        if (group.setDestroyed()) {
+            logger.info(groupId + " is destroyed.");
+            sendDestroyRaftNodeOps(group);
+        } else if (logger.isFineEnabled()) {
+            logger.fine(groupId + " is already destroyed.");
+        }
+    }
+
+    public void forceDestroyRaftGroup(String groupName) {
+        checkNotNull(groupName);
+        checkFalse(metadataGroupId.name().equalsIgnoreCase(groupName), "Cannot force-destroy the METADATA CP group!");
```

Comment:
Should we additionally check if we are not destroying the default group?

Reply:
`default` group can be destroyed too. If it loses its majority, users should force-destroy it.